flink-connector-bigquery-common/src/main/java/com/google/cloud/flink/bigquery/common/utils/AvroToBigQuerySchemaTransform.java

```diff
@@ -0,0 +1,456 @@
+            fields.add(
+                    convertAvroFieldToBigQueryField(
+                            nestedAvroSchema,
+                            avroSchema.getFields().get(i).name(),
+                            nestedLevel + 1));
+        }
+        FieldList nestedBigQueryFields = FieldList.of(fields);
+        return Field.newBuilder(name, LegacySQLTypeName.RECORD, nestedBigQueryFields)
+                .setDescription((avroSchema.getDoc() != null) ? avroSchema.getDoc() : null)
+                .setMode(Field.Mode.REQUIRED)
+                .build();
+    }
+
+    /**
+     * Helper function to convert the UNION type schema field to a BigQuery Field.
+     *
+     * <p>BigQuery supports nullable fields but not unions with multiple non-null types. This method
+     * validates the Avro UNION schema to ensure it conforms to BigQuery's requirements.
+     *
+     * <p>Valid UNION schemas are:
+     *
+     * <ul>
+     *   <li>["null", datatype] - Represents a nullable field.
+     *   <li>["datatype", null] - Represents a nullable field.
+     *   <li>[datatype] - Represents a non-nullable field.
+     * </ul>
+     *
+     * <p>Invalid UNION schemas include:
+     *
+     * <ul>
+     *   <li>["null"] - Only null type is not supported.
+     *   <li>[datatype1, datatype2, ...] - Unions without a null type are not supported.
+     *   <li>["null", ARRAY] Arrays in BigQuery cannot be nullable
+     *   <li>[ARRAY, "null"] Arrays in BigQuery cannot be nullable
+     * </ul>
+     *
+     * <p>If the UNION schema is valid, this method returns a BigQuery field with the schema of the
+     * non-null datatype. Otherwise, it throws an {@link UnsupportedOperationException}.
+     *
+     * @param avroSchema The Avro UNION schema to process.
+     * @return The BigQuery field corresponding to the non-null type in the UNION.
+     * @throws UnsupportedOperationException if the UNION schema is invalid for BigQuery.
+     */
+    private static Field convertAvroNullableFieldToBigQueryField(
+            org.apache.avro.Schema avroSchema, String name, int nestedLevel)
+            throws UnsupportedOperationException {
+        List<org.apache.avro.Schema> unionTypes = avroSchema.getTypes();
+
+        // Case, when there is only a single type in UNION.
+        // Can be ['null'] - ERROR
+        // [Valid-Datatype] - Not Nullable, element type
+        // Then it is essentially the same as not having a UNION.
+        if (unionTypes.size() == 1
+                && unionTypes.get(0).getType() != org.apache.avro.Schema.Type.NULL) {
+            return convertAvroFieldToBigQueryField(unionTypes.get(0), name, nestedLevel);
+        }
+
+        if (unionTypes.size() == 2) {
+            // Extract all the nonNull Datatypes.
+            // ['datatype, 'null'] and ['null', datatype] are only valid cases.
+            @Nullable org.apache.avro.Schema actualSchema = null;
```

Comment:
Dont need nullable annotation. Also, dont need explicit null initialization. Finally, `actualDataType` would be a better variable name.

Reply:
Yes, modified.